From: Daniel Stenberg <daniel@haxx.se>
Date: Sun, 23 Aug 2026 17:03:06 +0200
Subject: cookie: allow loading cookies for localhost

The added code to avoid loading PSL cookies from files from PR #22500
(c041895, not shipped in a release), calls psl_is_public_suffix(),
returns TRUE also for "localhost" (!).

We now allow cookies again for localhost. Added test 320 to verify.

Reported-by: Athos Ribeiro
Fixes #22645
Closes #22646

Origin: upstream, https://github.com/curl/curl/commit/85c8f6f3a5b107ee605122019cea4fb4edcfe61d
Last-Update: 2026-08-24
---
 lib/cookie.c                    | 12 ++++++++----
 tests/data/Makefile.am          |  2 +-
 tests/data/{test409 => test320} |  9 ++++-----
 3 files changed, 13 insertions(+), 10 deletions(-)
 copy tests/data/{test409 => test320} (74%)

diff --git a/lib/cookie.c b/lib/cookie.c
index c1a28bf..0b278e7 100644
--- a/lib/cookie.c
+++ b/lib/cookie.c
@@ -839,10 +839,14 @@ static bool is_public_suffix(struct Curl_easy *data,
           lcase[dlen] = 0;
           acceptable = psl_is_cookie_domain_acceptable(psl, lcase, lcookie);
         }
-        else
-          /* note that this PSL function returns the opposite value than
-             psl_is_cookie_domain_acceptable() does */
-          acceptable = !psl_is_public_suffix(psl, lcookie);
+        else {
+          /* libpsl says localhost is a PSL, we think not */
+          acceptable =
+            curl_strequal(lcookie, "localhost") ||
+            /* note that this PSL function returns the opposite value than
+               psl_is_cookie_domain_acceptable() does */
+            !psl_is_public_suffix(psl, lcookie);
+        }
         Curl_psl_release(data);
       }
       else
diff --git a/tests/data/Makefile.am b/tests/data/Makefile.am
index 5291ccc..5eb14c8 100644
--- a/tests/data/Makefile.am
+++ b/tests/data/Makefile.am
@@ -60,7 +60,7 @@ test280 test281 test282 test283 test284 test285 test286 test287 test288 \
 test289 test290 test291 test292 test293 test294 test295 test296 test297 \
 test298 test299 test300 test301 test302 test303 test304 test305 test306 \
 test307 test308 test309 test310 test311 test312 test313 test314 test315 \
-test316 test317 test318 test319                                         \
+test316 test317 test318 test319 test320                                 \
 test325 test326 test327 test328 test329 test330 test331 test332 test333 \
 test334 test335 test336 test337 test338 test339 test340 test341 test342 \
 test343 test344 test345 test346 test347 test348 test349 test350 test351 \
diff --git a/tests/data/test409 b/tests/data/test320
similarity index 74%
copy from tests/data/test409
copy to tests/data/test320
index f9a7131..645299d 100644
--- a/tests/data/test409
+++ b/tests/data/test320
@@ -35,14 +35,13 @@ PSL
 http
 </server>
 <name>
-Load cookies from file where some are PSL domains
+Load cookies from file for localhost
 </name>
 <command>
-http://foo.co.uk:%HTTPPORT/%TESTNUMBER -b %LOGDIR/cookies.txt --resolve foo.co.uk:%HTTPPORT:%HOSTIP
+http://localhost:%HTTPPORT/%TESTNUMBER -b %LOGDIR/cookies.txt --resolve localhost:%HTTPPORT:%HOSTIP
 </command>
 <file name="%LOGDIR/cookies.txt">
-.co.uk%TABTRUE%TAB/%TABFALSE%TAB0%TABbad%TABnot-allowed
-foo.co.uk%TABTRUE%TAB/%TABFALSE%TAB0%TABgood%TABallowed
+localhost%TABTRUE%TAB/%TABFALSE%TAB0%TABgood%TABallowed
 </file>
 </client>
 
@@ -50,7 +49,7 @@ foo.co.uk%TABTRUE%TAB/%TABFALSE%TAB0%TABgood%TABallowed
 <verify>
 <protocol crlf="headers">
 GET /%TESTNUMBER HTTP/1.1
-Host: foo.co.uk:%HTTPPORT
+Host: localhost:%HTTPPORT
 User-Agent: curl/%VERSION
 Accept: */*
 Cookie: good=allowed
